Ankit/ga4 web new

rvadera12 opened the pull request:
<!-- Hello and thank you for contributing to Segment action-destinations! -->

<!-- Before opening your pull request, make sure you have added and ran unit
     tests and tested your change locally. Refer to our testing
     documentation for more information: https://github.com/segmentio/action-destinations/blob/main/docs/testing.md -->

<!-- If you have questions or issues please open a new issue or create a new discussion
     post in Github. -->

_A summary of your pull request, including the what change you're making and why._

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [ ] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
- [ ] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)
- [ ] [Segmenters] Tested in the staging environment

SyedWasiHaider commented on packages/browser-destinations/src/destinations/index.ts:
```diff
@@ -45,7 +45,4 @@ register('635ada35ce269dbe305203ff', './logrocket')
 register('6372e18fb2b3d5d741c34bb6', './sabil')
 register('6372e1e36d9c2181f3900834', './wisepops')
 register('637c192eba61b944e08ee158', './vwo')
-register('638f843c4520d424f63c9e51', './commandbar')
```
Is this intentional?

rvadera12 replied:
Nope, just added back `commandbar` and `ripe`

zikaari commented on packages/browser-destinations/src/destinations/google-analytics-4-web/setConfigurationFields/index.ts:
```diff
@@ -0,0 +1,148 @@
+    }
+    if (payload.page_title) {
+      gtag('set', { page_title: payload.page_title })
+    }
+    if (payload.page_referrer) {
+      gtag('set', { page_referrer: payload.page_referrer })
+    }
+    if (payload.page_location) {
+      gtag('set', { page_location: payload.page_location })
+    }
+    if (payload.language) {
+      gtag('set', { language: payload.language })
+    }
+    if (payload.content_group) {
+      gtag('set', { content_group: payload.content_group })
+    }
+    if (payload.campaign_term) {
+      gtag('set', { campaign_term: payload.campaign_term })
+    }
+    if (payload.campaign_source) {
+      gtag('set', { campaign_source: payload.campaign_source })
+    }
+    if (payload.campaign_name) {
+      gtag('set', { campaign_name: payload.campaign_name })
+    }
+    if (payload.campaign_medium) {
+      gtag('set', { campaign_medium: payload.campaign_medium })
+    }
+    if (payload.campaign_id) {
+      gtag('set', { campaign_id: payload.campaign_id })
+    }
+    if (payload.campaign_content) {
+      gtag('set', { campaign_content: payload.campaign_content })
+    }
+    if (payload.user_id) {
+      gtag('set', { user_id: payload.user_id })
+    }
+    if (payload.user_properties) {
+      gtag('set', { user_properties: payload.user_properties })
+    }
```
```suggestion
    const properties = [
      'screen_resolution',
      'page_title',
      'page_referrer',
      '...'
    ]

    gtag('set', Object.fromEntries(
      properties
        .map((key) => [key, payload[key]])
        .filter(([, value]) => !value)
    ))
```

zikaari replied:
The `properties` key array might as well be type decorated assuming that type is available

silesky replied:
do we have a polyfill for fromEntries? I think we might want to use reduce instead...

zikaari commented on packages/browser-destinations/src/destinations/google-analytics-4-web/types.ts:
```diff
@@ -0,0 +1,4 @@
+export type GA = {
+    //dataLayer: any
```
Should we just remove it?

zikaari reviewed (commented):
Nice work!

Can someone help me understand why the tests create a "local" subscription model instead of testing against the "real" default subscription that the users are likely going to be facing?

Also, I'm not seeing any event filtering _(based on the `event` property)_ across the actions' `defaultSubscription` that'd discern one action from another at runtime, as seen [here ](https://github.com/segmentio/action-destinations/blob/main/packages/browser-destinations/src/destinations/braze/trackPurchase/index.ts#L9). How does the destination express which partner should be triggered out of the many available?

pooyaj commented on packages/browser-destinations/src/destinations/google-analytics-4-web/index.ts:
```diff
@@ -0,0 +1,199 @@
+    }
+    const script = `https://www.googletagmanager.com/gtag/js?id=${settings.measurementID}`
+    await deps.loadScript(script)
+    return window.gtag
```
When `loadScript` completes, does it replace `window.gtag` with a different object? If so, we may need to make sure that `window.gtag` points to the full `gtag` library rather than the stub using `await deps.resolveWhen ...`. Might not be necessary, but worth double checking

pooyaj reviewed (approved):
Looks great 🙌  Only one minor comment to verify.

pooyaj commented on packages/browser-destinations/src/destinations/google-analytics-4-web/__tests__/addPaymentInfo.test.ts:
```diff
@@ -0,0 +1,86 @@
+      ...setting,
+      subscriptions
+    })
+    jest.spyOn(destination, 'initialize')
```
In all of the tests can we do this instead: 
```
jest.spyOn(destination, 'initialize').mockImplementation(() => {
... // see examples of these in other web destinations
})
```

Because currently we are running the initialize method, and actually trying to load the GA tag from Google in each run of each test 😬 Makes tests flaky, and they fail with no network connects!

rvadera12 commented on packages/browser-destinations/src/destinations/google-analytics-4-web/index.ts:
```diff
@@ -0,0 +1,199 @@
+      allow_google_signals: settings.allowGoogleSignals
+    }
+
+    window.dataLayer = window.dataLayer || []
```
Going to leave this the same so it matches exactly to google's implementation.

rvadera12 commented:
Thanks for reviewing the PR @zikaari! 
> 
> Can someone help me understand why the tests create a "local" subscription model instead of testing against the "real" default subscription that the users are likely going to be facing?
>
Good point! There is not specific reason, the mappings actually do all match up for the most part. However we were using the google analytics 4 cloud mode tests as a framework and implemented the tests in the same method. I will make a ticket to fix this, but will move forward without changing it in this PR. 

> Also, I'm not seeing any event filtering _(based on the `event` property)_ across the actions' `defaultSubscription` that'd discern one action from another at runtime, as seen [here ](https://github.com/segmentio/action-destinations/blob/main/packages/browser-destinations/src/destinations/braze/trackPurchase/index.ts#L9). How does the destination express which partner action should be triggered out of the many available?

Great catch, I updated the event properties for the default subscriptions, for all of the actions that have a corresponding event name that matches the Segment Spec!
